### PR TITLE
Add support for tslint version to 5.1.0

### LINKED ISSUE #354: tslint 5.0.0 contains custom rule breaking changes
Since [tslint v5.0.0](https://github.com/palantir/tslint/releases/tag/5.0.0), there is a breaking change to the API for custom tslint rules. 

The good news is that it looks like changing this is a pretty straightforward refactoring. 

```diff
- public applyWithProgram(srcFile: ts.SourceFile, langSvc: ts.LanguageService): Lint.RuleFailure[] {
-     return this.applyWithWalker(new Walker(srcFile, this.getOptions(), langSvc.getProgram()));
+ public applyWithProgram(srcFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+     return this.applyWithWalker(new Walker(srcFile, this.getOptions(), program));
```

However changing this would also mean a breaking change to this package/ruleset, since it would make tslint versions >= 5.0.0 the new dependency. Are there any concerns with that? Do we need to ensure that any new updates to these rules are compatible with projects using older version of tslint?

### LINKED ISSUE #364: remove no-sparse-arrays rule
TSLint 5.1.0 was just released and it contains the `no-sparse-arrays` rule, so the rule by the same name in this project can be safely removed when support for TSLint 5.x is released.

http://palantir.github.io/tslint/rules/no-sparse-arrays

### PR DESCRIPTION
**Pull request by bgold09**
resolves #354 

* 'no-cookies' now explicitly require type checking (extends
 Lint.Rules.TypedRule)
* The following rules now optionally require type checking (extend
from Lint.Rules.OptionallyTypedRule):
  * no-string-based-set-immediate
  * no-string-based-set-interval
  * no-string-based-set-timeout
* 'no-single-line-block-comment' no longer supports nested single
line comments
* Removed 'no-unused-imports' in favor of the TSLint implementation
* The default severity for TSLint is now error; fixing easy violations that
were exposed by that change
* Add extended metadata for new rules provided by TSLint
* Remove deprecated rule 'no-unused-imports'
* Remove deprecated rule no-sparse-arrays (fixes #364)

### CODE REVIEW

File: src/noForInRule.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export class Rule extends Lint.Rules.AbstractRule {
     public static FAILURE_STRING_FACTORY(initializer: string, expression: string): string {
         //tslint:disable-next-line:max-line-length
         return `Do not use the 'for in' statement: 'for (${initializer} in ${expression})'. If this is an object, use 'Object.keys' instead. If this is an array use a standard 'for' loop instead.`;
-    };
+    }
```
Comment (HamletDRC) [inline src/noForInRule.ts]: I see that tslint still has a bug in their unnecessary-semicolon rule

File: src/noInvalidRegexpRule.ts [hunk 2]
```diff
@@ -46,9 +46,8 @@ class NoInvalidRegexpRuleWalker extends ErrorTolerantWalker {
                 if (arg1.kind === ts.SyntaxKind.StringLiteral) {
                     const regexpText: string = (<ts.StringLiteral>arg1).text;
                     try {
-                        /* tslint:disable:no-unused-new */
+                        // tslint:disable-next-line:no-unused-expression
```
Comment (HamletDRC) [inline src/noInvalidRegexpRule.ts]: +1

**Comment by HamletDRC** [comment]
This all looks good and I ran it on MSE and there were no issues or false positives (however, we are not doing "full program" scanning)

**Comment by HamletDRC** [comment]
If I upgrade to TypeScript 2.3 then some tests fail though :(

**Comment by HamletDRC** [comment]
Pushed and closed!

**Comment by sapegin** [comment]
@HamletDRC Could you please publish 5.0.0 to npm?

**Comment by HamletDRC** [comment]
@sapegin I started the process now, thanks.

**Pull request closed without merging**